Source/WebCore/dom/Document.cpp

```diff
@@ -3942,28 +3945,90 @@ ExceptionOr<void> Document::write(Document* entryDocument, SegmentedString&& tex
     return { };
 }
 
-ExceptionOr<void> Document::write(Document* entryDocument, FixedVector<String>&& strings)
+ExceptionOr<void> Document::write(Document* entryDocument, FixedVector<std::variant<RefPtr<TrustedHTML>, String>>&& strings)
 {
-    if (!isHTMLDocument() || m_throwOnDynamicMarkupInsertionCount)
-        return Exception { ExceptionCode::InvalidStateError };
+    auto isTrusted = true;
+    SegmentedString text;
+    for (auto& entry : strings) {
+        text.append(
+            WTF::switchOn(
+                WTFMove(entry),
+                [&isTrusted](const String& string) -> String {
+                    isTrusted = false;
+                    return string;
+                },
+                [](const RefPtr<TrustedHTML>& html) -> String {
+                    return html->toString();
+                }
+            )
+        );
+    }
+
+    if (isTrusted || !scriptExecutionContext()->settingsValues().trustedTypesEnabled)
+        return write(entryDocument, WTFMove(text));
+
+    String textString = String(text.toString());
+    auto stringValueHolder = trustedTypeCompliantString(TrustedType::TrustedHTML, *scriptExecutionContext(), textString, "Document write"_s);
+    if (stringValueHolder.hasException())
+        return stringValueHolder.releaseException();
+    text.clear();
+    text.close();
+    SegmentedString trustedSegmentedText(stringValueHolder.releaseReturnValue());
+    return write(entryDocument, WTFMove(trustedSegmentedText));
+}
 
+ExceptionOr<void> Document::write(Document* entryDocument, FixedVector<String>&& strings)
+{
     SegmentedString text;
     for (auto& string : strings)
         text.append(WTFMove(string));
 
     return write(entryDocument, WTFMove(text));
 }
 
-ExceptionOr<void> Document::writeln(Document* entryDocument, FixedVector<String>&& strings)
+ExceptionOr<void> Document::writeln(Document* entryDocument, FixedVector<std::variant<RefPtr<TrustedHTML>, String>>&& strings)
 {
-    if (!isHTMLDocument() || m_throwOnDynamicMarkupInsertionCount)
-        return Exception { ExceptionCode::InvalidStateError };
+    auto isTrusted = true;
+    SegmentedString text;
```

Comment:
It’s not OK to repeat so much code in `write` and `writeln`; copy and paste code is not a good pattern. We need to find a way to share code.

Reply:
Okay I'll make a shared function that they both call into, I did consider it but then thought well they were separated to start with. But you're right it is too repetitive.